docs/metrics.md

```diff
@@ -0,0 +1,561 @@
+counter.add(3, &[KeyValue::new("color", "green"), KeyValue::new("name", "lime")]);
+counter.add(5, &[KeyValue::new("color", "yellow"), KeyValue::new("name", "lemon")]);
+counter.add(8, &[KeyValue::new("name", "lemon"), KeyValue::new("color", "yellow")]); // bad perf
+```
+
+:heavy_check_mark: You should use consistent ordering of attributes to achieve
+the best performance.
+
+:heavy_check_mark: If feasible, provide the attributes in the sorted order of
+`Key`s, to minimize memory usage within the Metrics SDK.
+
+```rust
+let counter = meter.u64_counter("example_counter").build();
+counter.add(2, &[KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+```
+
+### Reporting measurements via synchronous instruments
+
+:heavy_check_mark:  When reporting measurements, use array slices for attributes
+rather than creating vectors. Arrays are more efficient as they avoid
+unnecessary heap allocations on the measurement path:
+
+```rust
+// Good practice: Using an array slice directly
+counter.add(2, &[KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+
+// Avoid this: Creating a Vec is unnecessary, and it allocates on the heap
+// counter.add(2, &vec![KeyValue::new("color", "red"), KeyValue::new("name", "apple")]);
+```
+
+This approach minimizes memory allocations on metrics hot paths, improving
+overall performance when reporting metrics frequently.
+
+### Reporting measurements via asynchronous instruments
+
+Asynchronous instruments like `ObservableCounter` are ideal for reporting
+metrics that are already being tracked or stored elsewhere in your application.
+These instruments allow you to observe and report the current state of existing
+values without manually managing the aggregation yourself.
+
+:heavy_check_mark: Use `ObservableCounter` when you already have a variable
+tracking a count:
+
+```rust
+// Example: Using ObservableCounter when you already have a variable tracking counts
+use opentelemetry::KeyValue;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+// An existing variable in your application
+static REQUEST_COUNT: AtomicU64 = AtomicU64::new(0);
+
+// In your application code, you update this directly
+fn handle_request() {
+    // Process request...
+    REQUEST_COUNT.fetch_add(1, Ordering::SeqCst);
+}
+
+// When setting up metrics, register an observable counter that reads from your variable
+fn setup_metrics(meter: &opentelemetry::metrics::Meter) {
+    let _observable_counter = meter
+        .u64_observable_counter("request_count")
+        .with_description("Number of requests processed")
+        .with_unit("requests")
+        .with_callback(|observer| {
+            // Read the current value from your existing counter
+            observer.observe(
+                REQUEST_COUNT.load(Ordering::SeqCst), 
+                &[KeyValue::new("endpoint", "/api")]
+            )
+        })
+        .build();
+}
+```
+
+:heavy_check_mark: Use regular Counter when you need to increment counts at
```

Comment:
```suggestion
:heavy_check_mark: Use synchronous Counter when you need to increment counts at
```

Reply:
Also, the example for the regular Counter seems incorrectly placed under the asynchronous instruments section, which could cause significant confusion for readers. Probably, we can expand the sync instrument section by adding this example and add a compare/contrast note here.